api/document/serializers/util.py

```diff
@@ -58,3 +59,25 @@ def list_to_internal_value(data: List[PrimitiveDict],
     # want.
     serializer = field_class(*args, **{'many': True, **kwargs})
     return serializer.to_internal_value(data)
+
+
+class SourcelineErrorMixin:
+    '''
+    A mixin for DRF Serializers that adds metadata about the
+    source line number that validation errors originated
+    from, if possible.
+
+    In order for this to work, the data passed to the
+    Serializer must have source line number metadata.
+    '''
+
+    def run_validation(self, data=empty):
+        try:
+            return super().run_validation(data)  # type: ignore
```

Comment:
I had to tell mypy to ignore this, as otherwise it would complain that this class' superclass has no `run_validation()` method.

Hmm, maybe mixins actually aren't the right approach here. I could turn this into a context manager instead, and then just add a trivial `run_validation()` to the two classes that need it, which simply wraps `super().run_validation()` in the context manager.  Or for that matter I could make it a decorator...

Reply:
I like that context managers would be a but more explicit about when the "magic" is happening, so I'd lean that direction.